VoodooPS2Trackpad/VoodooPS2ALPSGlidePoint.cpp

```diff
@@ -272,549 +415,3921 @@ bool ApplePS2ALPSGlidePoint::start( IOService * provider )
+        i = y_low.num_bits / 2;
+        y_low.num_bits = y_low.num_bits - i;
+        y_high.start_bit = y_low.start_bit + i;
+        y_high.num_bits = max(i, 1);
+    }
+    
+    /* top-left corner */
+    corner[0].x = (priv->x_max * (2 * x_low.start_bit + x_low.num_bits - 1)) /
+    (2 * (priv->x_bits - 1));
+    corner[0].y = (priv->y_max * (2 * y_low.start_bit + y_low.num_bits - 1)) /
+    (2 * (priv->y_bits - 1));
+    
+    /* top-right corner */
+    corner[1].x = (priv->x_max * (2 * x_high.start_bit + x_high.num_bits - 1)) /
+    (2 * (priv->x_bits - 1));
+    corner[1].y = (priv->y_max * (2 * y_low.start_bit + y_low.num_bits - 1)) /
+    (2 * (priv->y_bits - 1));
+    
+    /* bottom-right corner */
+    corner[2].x = (priv->x_max * (2 * x_high.start_bit + x_high.num_bits - 1)) /
+    (2 * (priv->x_bits - 1));
+    corner[2].y = (priv->y_max * (2 * y_high.start_bit + y_high.num_bits - 1)) /
+    (2 * (priv->y_bits - 1));
+    
+    /* bottom-left corner */
+    corner[3].x = (priv->x_max * (2 * x_low.start_bit + x_low.num_bits - 1)) /
+    (2 * (priv->x_bits - 1));
+    corner[3].y = (priv->y_max * (2 * y_high.start_bit + y_high.num_bits - 1)) /
+    (2 * (priv->y_bits - 1));
+    
+    /* x-bitmap order is reversed on v5 touchpads  */
+    if (priv->proto_version == ALPS_PROTO_V5) {
+        for (i = 0; i < 4; i++)
+            corner[i].x = priv->x_max - corner[i].x;
+    }
+    
+    /* y-bitmap order is reversed on v3 and v4 touchpads  */
+    if (priv->proto_version == ALPS_PROTO_V3 || priv->proto_version == ALPS_PROTO_V4) {
+        for (i = 0; i < 4; i++)
+            corner[i].y = priv->y_max - corner[i].y;
+    }
+    
+    /*
+     * We only select a corner for the second touch once per 2 finger
+     * touch sequence to avoid the chosen corner (and thus the coordinates)
+     * jumping around when the first touch is in the middle.
+     */
+    if (priv->second_touch == -1) {
+        /* Find corner closest to our st coordinates */
+        closest = 0x7fffffff;
+        for (i = 0; i < 4; i++) {
+            int dx = fields->st.x - corner[i].x;
+            int dy = fields->st.y - corner[i].y;
+            int distance = dx * dx + dy * dy;
+            
+            if (distance < closest) {
+                priv->second_touch = i;
+                closest = distance;
+            }
+        }
+        /* And select the opposite corner to use for the 2nd touch */
+        priv->second_touch = (priv->second_touch + 2) % 4;
+    }
+    
+    fields->mt[0] = fields->st;
+    fields->mt[1] = corner[priv->second_touch];
+    
+#if DEBUG
+    IOLog("ALPS: BITMAP\n");
+    
+    unsigned int ymap = fields->y_map;
+    
+    for (int i = 0; ymap != 0; i++, ymap >>= 1) {
+        unsigned int xmap = fields->x_map;
+        char bitLog[160];
+        strlcpy(bitLog, "ALPS: ", sizeof("ALPS: ") + 1);
+        
+        for (int j = 0; xmap != 0; j++, xmap >>= 1) {
+            strcat(bitLog, (ymap & 1 && xmap & 1) ? "1 " : "0 ");
```

Comment:
Use strlcat.

Reply:
Changed. Can you check if size_t is correctly used?